spinalcordtoolbox/scripts/sct_maths.py

```diff
@@ -270,223 +315,465 @@ def __call__(self, _, namespace, values, option_string=None):
-    dim_list = ['x', 'y', 'z', 't']
+class SctMathsTypeError(Exception):
+    """Inappropriate argument type for an `sct_maths` operation."""
 
-    if arg_name == "otsu":
-        data_out = sct_math.otsu(data, arg_value)
 
-    elif arg_name == "adap":
-        data_out = sct_math.adap(data, arg_value[0], arg_value[1])
+def check_image_to_image(current_type: str, *args, **kwargs) -> str:
+    """Type checker for generic image -> image operations."""
+    if current_type != 'image':
+        raise SctMathsTypeError(f"expected image, but got {current_type}")
+    return 'image'
 
-    elif arg_name == "otsu_median":
-        data_out = sct_math.otsu_median(data, arg_value[0], arg_value[1])
 
-    elif arg_name == "thr":
-        data_out = sct_math.threshold(data, lthr=arg_value)
+def check_metric(current_type: str, fname) -> str:
+    """Type checker for similarity metrics."""
+    if current_type != 'image':
+        raise SctMathsTypeError(f"expected image, but got {current_type}")
+    return 'metric'
 
-    elif arg_name == "uthr":
-        data_out = sct_math.threshold(data, uthr=arg_value)
 
-    elif arg_name == "percent":
-        data_out = sct_math.perc(data, arg_value)
+def check_arithmetic(current_type: str, *args) -> str:
+    """Type checker for `-add`, `-sub`, etc."""
+    # special case for -add, -mul to operate on a 4D volume across the t axis
+    if not args:
+        return check_image_to_image(current_type)
 
-    elif arg_name == "bin":
-        data_out = sct_math.binarize(data, bin_thr=arg_value)
+    # in the usual case, numbers can get promoted to images
+    if current_type == 'number':
+        if all(isinstance(arg, float) for arg in args):
+            return 'number'
+        else:
+            return 'image'
+    elif current_type == 'image':
+        return 'image'
+    else:
+        raise SctMathsTypeError(f"expected number or image, but got {current_type}")
+
+
+# the type-checking function for each sct_maths operation
+CHECK = {
+    "add": check_arithmetic,
+    "sub": check_arithmetic,
+    "mul": check_arithmetic,
+    "div": check_arithmetic,
+    "mean": check_image_to_image,
+    "rms": check_image_to_image,
+    "std": check_image_to_image,
+    "bin": check_image_to_image,
+    "otsu": check_image_to_image,
+    "adap": check_image_to_image,
+    "otsu_median": check_image_to_image,
+    "percent": check_image_to_image,
+    "thr": check_image_to_image,
+    "uthr": check_image_to_image,
+    "dilate": check_image_to_image,
+    "erode": check_image_to_image,
+    "smooth": check_image_to_image,
+    "laplacian": check_image_to_image,
+    "denoise": check_image_to_image,
+    "mi": check_metric,
+    "minorm": check_metric,
+    "corr": check_metric,
+    "symmetrize": check_image_to_image,
+}
+
+
+class SctMathsValueError(Exception):
+    """Inappropriate argument value for an `sct_maths` operation."""
+
+
+def get_header_and_values(
```

Comment:
I took me a while to come up with this argument normalizer, but I'm happy with it. Making sure that it had a somewhat reasonable type signature really helped me figure it out.